### PR TITLE
Refactoring validation middleware

### PR DESCRIPTION
**Pull request by Tirokk**
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-10T15:37:53Z" title="Friday, May 10th 2019, 5:37:53 pm +02:00">May 10, 2019</time>_
_Merged <time datetime="2019-05-14T10:03:00Z" title="Tuesday, May 14th 2019, 12:03:00 pm +02:00">May 14, 2019</time>_
---

## Pullrequest
Two things:

* Fix `UpdateUser` test block (was not testing anything).
* Put all validations in one place

@mattwr18 @ulfgebhardt please read my commit messages in this PR.
### Issues
This is a follow-up on #478 

### Checklist

- [X] None

### How2Test

- [X] None

### Todo

- [X] None

### CODE REVIEW

File: backend/src/resolvers/users.spec.js [hunk 1]
```diff
@@ -5,15 +5,15 @@ import Factory from '../seed/factories'
 const factory = Factory()
 let client
 
-afterAll(async () => {
+afterEach(async () => {
```
Comment (Tirokk) [inline backend/src/resolvers/users.spec.js:8]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-10T15:39:59Z" title="Friday, May 10th 2019, 5:39:59 pm +02:00">May 10, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -5,15 +5,15 @@ import Factory from '../seed/factories'
 const factory = Factory()
 let client
 
-afterAll(async () => {
+afterEach(async () => {
```


@ulfgebhardt `afterEach` means: Run this after each test case. `afterAll` means: Run this after the entire test **suite**. `afterAll` is great for performance tweaks, but you have to tear down the created data yourself. Read: https://jestjs.io/docs/en/setup-teardown

File: backend/src/middleware/validation/index.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
+  } else {
+    throw new UserInputError('Input is not a URL')
+  }
+}
+
+export default {
+  Mutation: {
+    CreateUser: validateUsername,
```
Comment (Tirokk) [inline backend/src/middleware/validation/index.js:27]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-10T15:41:47Z" title="Friday, May 10th 2019, 5:41:47 pm +02:00">May 10, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,31 @@
+import { UserInputError } from 'apollo-server'
+
+const USERNAME_MIN_LENGTH = 3
+
+const validateUsername = async (resolve, root, args, context, info) => {
+  if (args.name && args.name.length >= USERNAME_MIN_LENGTH) {
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
+  } else {
+    throw new UserInputError(`Username must be at least ${USERNAME_MIN_LENGTH} characters long!`)
+  }
+}
+
+const validateUrl = async (resolve, root, args, context, info) => {
+  const { url } = args
+  const isValid = url.match(/^(?:https?:\/\/)(?:[^@\n])?(?:www\.)?([^:/\n?]+)/g)
+  if (isValid) {
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
+  } else {
+    throw new UserInputError('Input is not a URL')
+  }
+}
+
+export default {
+  Mutation: {
+    CreateUser: validateUsername,
```


@ulfgebhardt we should validate `username` also on creation, no? Without it, the validation would be kind of pointless.

... Except, the `CreateUser` will be called on registration and we supply no username at all. But we'll see. For now, I think it's safer to have it checked on Create+Update.

File: backend/src/resolvers/users.spec.js [hunk 3]
```diff
@@ -35,34 +36,39 @@ describe('userMiddleware', () => {
     })
   })
 
-  describe('update User', () => {
+  describe('UpdateUser', () => {
+    beforeEach(async () => {
+      await factory.create('User', { id: 'u47', name: 'John Doe' })
```
Comment (Tirokk) [inline backend/src/resolvers/users.spec.js:41]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-10T15:42:22Z" title="Friday, May 10th 2019, 5:42:22 pm +02:00">May 10, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -35,34 +36,39 @@ describe('userMiddleware', () => {
     })
   })
 
-  describe('update User', () => {
+  describe('UpdateUser', () => {
+    beforeEach(async () => {
+      await factory.create('User', { id: 'u47', name: 'John Doe' })
```


@ulfgebhardt you tried to update a user that did not exist.

**Comment by Tirokk** [comment]
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-10T16:00:43Z" title="Friday, May 10th 2019, 6:00:43 pm +02:00">May 10, 2019</time>_

---


Me: Adding/fixing tests
Codecov: Test coverage dropped by 0.01%.

:laughing:

File: backend/src/middleware/validation/index.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import { UserInputError } from 'apollo-server'
+
+const USERNAME_MIN_LENGTH = 3
+
+const validateUsername = async (resolve, root, args, context, info) => {
+  if (args.name && args.name.length >= USERNAME_MIN_LENGTH) {
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
```
Comment (Tirokk) [inline backend/src/middleware/validation/index.js:8]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-05-10T17:18:54Z" title="Friday, May 10th 2019, 7:18:54 pm +02:00">May 10, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,31 @@
+import { UserInputError } from 'apollo-server'
+
+const USERNAME_MIN_LENGTH = 3
+
+const validateUsername = async (resolve, root, args, context, info) => {
+  if (args.name && args.name.length >= USERNAME_MIN_LENGTH) {
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
```


are we updating the user twice here? once in the validation middleware and once on line 15 of the `userMiddleware`?

File: backend/src/middleware/validation/index.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    throw new UserInputError(`Username must be at least ${USERNAME_MIN_LENGTH} characters long!`)
+  }
+}
+
+const validateUrl = async (resolve, root, args, context, info) => {
+  const { url } = args
+  const isValid = url.match(/^(?:https?:\/\/)(?:[^@\n])?(?:www\.)?([^:/\n?]+)/g)
+  if (isValid) {
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
```
Comment (Tirokk) [inline backend/src/middleware/validation/index.js:19]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-05-10T17:28:07Z" title="Friday, May 10th 2019, 7:28:07 pm +02:00">May 10, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,31 @@
+import { UserInputError } from 'apollo-server'
+
+const USERNAME_MIN_LENGTH = 3
+
+const validateUsername = async (resolve, root, args, context, info) => {
+  if (args.name && args.name.length >= USERNAME_MIN_LENGTH) {
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
+  } else {
+    throw new UserInputError(`Username must be at least ${USERNAME_MIN_LENGTH} characters long!`)
+  }
+}
+
+const validateUrl = async (resolve, root, args, context, info) => {
+  const { url } = args
+  const isValid = url.match(/^(?:https?:\/\/)(?:[^@\n])?(?:www\.)?([^:/\n?]+)/g)
+  if (isValid) {
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
```


same as :point_up:

File: backend/src/middleware/validation/index.js (hunk 4 above)
Comment (Tirokk) [reply backend/src/middleware/validation/index.js:8]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-10T17:58:02Z" title="Friday, May 10th 2019, 7:58:02 pm +02:00">May 10, 2019</time>_

---


No, calling `resolve` will continue the chain of middlewares until you hit the final resolver.

File: backend/src/resolvers/users.spec.js [hunk 6]
```diff
@@ -71,7 +77,8 @@ describe('userMiddleware', () => {
 
     it('with too short name', async () => {
       const variables = {
-        id: 'u1'
+        id: 'u47',
```
Comment (Tirokk) [inline backend/src/resolvers/users.spec.js:80]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-10T17:59:52Z" title="Friday, May 10th 2019, 7:59:52 pm +02:00">May 10, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -71,7 +77,8 @@ describe('userMiddleware', () => {
 
     it('with too short name', async () => {
       const variables = {
-        id: 'u1'
+        id: 'u47',
```


I think I have to give myself a "Request Changes" review here. I told @ulfgebhardt not to create test cases (==specification) for behaviour that we actually don't want to support, e.g. passing `id` from the client for `CreateUser` mutation. I'll update that as soon as possible. I think, at first, I thought I have to pass the `id` to make it happen. That's why it's there.

**Comment by Tirokk** [comment]
> [ulfgebhardt](https://github.com/ulfgebhardt) **Authored by [ulfgebhardt](https://github.com/ulfgebhardt)**
_<time datetime="2019-05-12T00:38:37Z" title="Sunday, May 12th 2019, 2:38:37 am +02:00">May 12, 2019</time>_

---


> Me: Adding/fixing tests
> Codecov: Test coverage dropped by 0.01%.
> 
> 😆

The Backend Test Setup needs to be fixed to comply with standard test runs of jest in oder to collect coverage properly.

**Comment by Tirokk** [comment]
> [codecov[bot]](https://github.com/apps/codecov) **Authored by [codecov[bot]](https://github.com/apps/codecov)**
_<time datetime="2019-05-13T11:03:52Z" title="Monday, May 13th 2019, 1:03:52 pm +02:00">May 13, 2019</time>_

---


# [Codecov](https://codecov.io/gh/Human-Connection/Human-Connection/pull/596?src=pr&el=h1) Report
> Merging [#596](https://codecov.io/gh/Human-Connection/Human-Connection/pull/596?src=pr&el=desc) into [master](https://codecov.io/gh/Human-Connection/Human-Connection/commit/fe386e235581faf06929604e4addac313bed0b99?src=pr&el=desc) will **increase** coverage by `0.38%`.
> The diff coverage is `0%`.

[Impacted file tree graph](https://codecov.io/gh/Human-Connection/Human-Connection/pull/596?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #596      +/-   ##
==========================================
+ Coverage   20.14%   20.52%   +0.38%     
==========================================
  Files         157      157              
  Lines        2210     2237      +27     
  Branches      105      113       +8     
==========================================
+ Hits          445      459      +14     
- Misses       1741     1752      +11     
- Partials       24       26       +2
```


| [Impacted Files](https://codecov.io/gh/Human-Connection/Human-Connection/pull/596?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [backend/src/middleware/index.js](https://codecov.io/gh/Human-Connection/Human-Connection/pull/596/diff?src=pr&el=tree#diff-YmFja2VuZC9zcmMvbWlkZGxld2FyZS9pbmRleC5qcw==) | `0% <ø> (ø)` | :arrow_up: |
| [backend/src/middleware/userMiddleware.js](https://codecov.io/gh/Human-Connection/Human-Connection/pull/596/diff?src=pr&el=tree#diff-YmFja2VuZC9zcmMvbWlkZGxld2FyZS91c2VyTWlkZGxld2FyZS5qcw==) | `0% <ø> (ø)` | :arrow_up: |
| [backend/src/middleware/validation/index.js](https://codecov.io/gh/Human-Connection/Human-Connection/pull/596/diff?src=pr&el=tree#diff-YmFja2VuZC9zcmMvbWlkZGxld2FyZS92YWxpZGF0aW9uL2luZGV4Lmpz) | `0% <0%> (ø)` | |
| [webapp/components/Modal.vue](https://codecov.io/gh/Human-Connection/Human-Connection/pull/596/diff?src=pr&el=tree#diff-d2ViYXBwL2NvbXBvbmVudHMvTW9kYWwudnVl) | `75% <0%> (-1.92%)` | :arrow_down: |
| [webapp/pages/logout.vue](https://codecov.io/gh/Human-Connection/Human-Connection/pull/596/diff?src=pr&el=tree#diff-d2ViYXBwL3BhZ2VzL2xvZ291dC52dWU=) | `0% <0%> (ø)` | :arrow_up: |
| [webapp/components/ContentMenu.vue](https://codecov.io/gh/Human-Connection/Human-Connection/pull/596/diff?src=pr&el=tree#diff-d2ViYXBwL2NvbXBvbmVudHMvQ29udGVudE1lbnUudnVl) | `39.47% <0%> (+3.47%)` | :arrow_up: |

**Comment by Tirokk** [comment]
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-13T12:20:23Z" title="Monday, May 13th 2019, 2:20:23 pm +02:00">May 13, 2019</time>_

---


@ulfgebhardt did you see this? Check commit messages and "Files Changed" tab. If yes, you can merge this PR.

**Review by Mogge — APPROVED** [review]
Nice @roschaefer, I was planning on extracting some validations out. 
I'm approving these changes and will create a separate issue to refactor the comments validations

**Pull request closed without merging**